cmd/gcp-controller-manager/node_csr_approver.go

```diff
@@ -657,40 +653,64 @@ func getClusterInstanceGroupUrls(ctx *controllerContext, instanceZone string) ([
+// Since this is user-modifiable, we should still verify membership.
+// However, we can avoid some GCE API ListManagedInstanceGroupInstances calls using the hint.
+
+// Verify that the instanceGroupHint is in fact
+// part of the cluster's known instance groups
+//
+// if it is: return the resolved instanceGroup
+// else: ""
+func validateInstanceGroupHint(instanceGroupUrls []string, instanceGroupHint string) (string, error) {
+	if instanceGroupHint == "" {
+		return "", fmt.Errorf("validateInstanceGroupHint: hint is empty")
 	}
 
-	// InstanceGroupHint is the name of the instancegroup obtained from the instance metadata.
-	// Since this is user-modifiable, we should still verify membership.
-	// However, we can avoid some GCE API ListManagedInstanceGroupInstances calls using the hint.
+	location, igName, err := parseInstanceGroupURL(instanceGroupHint)
+	if err != nil {
+		return "", err
+	}
 
-	// verify that the instanceGroupHint is in fact
-	// part of the cluster's known instance groups
-	var instanceGroupHintValid = false
-	var instanceGroupHintResolved = ""
-	for _, g := range clusterInstanceGroupUrls {
-		// perform a suffix match.
-		// instanceGroupHint is of the form project/p0/zones/z0/instanceGroup...
-		// g includes the hostname: https://www.googleapis.com/compute/v1/project/...
-		if strings.HasSuffix(g, instanceGroupHint) {
-			instanceGroupHintValid = true
-			instanceGroupHintResolved = g
+	var resolved string
+	for _, g := range instanceGroupUrls {
+		gl, gn, err := parseInstanceGroupURL(g)
+		if err != nil {
+			return "", err
+		}
+		if gl == location && gn == igName {
+			resolved = g
 		}
 	}
 
-	// If the instanceGroupHint points is verified to be part of the cluster,
-	// insert the hinted group at the beginning of the cluster instancegroup list
-	if instanceGroupHintValid {
-		clusterInstanceGroupUrls = append([]string{instanceGroupHintResolved}, clusterInstanceGroupUrls...)
-	} else if instanceGroupHint != "" {
-		// Not expected. This will lead to a linear search in all instance groups in the cluster.
-		klog.Infof("hinted instance group %q not found in cluster.", instanceGroupHint)
+	if resolved == "" {
+		return "", fmt.Errorf("hinted instance group %q not found in cluster", instanceGroupHint)
 	}
 
+	return resolved, nil
+}
+
+func clusterHasInstance(ctx *controllerContext, instanceZone string, instanceID uint64, instanceGroupHint string) (bool, error) {
+	clusterInstanceGroupUrls, err := getClusterInstanceGroupUrls(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	validatedInstanceGroupHint, err := validateInstanceGroupHint(clusterInstanceGroupUrls, instanceGroupHint)
+	if err != nil {
+		klog.Warningf("error validating instance group: %v", err)
+	} else {
+		clusterInstanceGroupUrls = append([]string{validatedInstanceGroupHint}, clusterInstanceGroupUrls...)
+	}
+
+	klog.V(3).Infof("Cluster %q clusterInstanceGroupUrls %+v", ctx.gcpCfg.ClusterName, clusterInstanceGroupUrls)
```

Comment:
Do we need cluster name in the log? gcp controller manager runs in the context of a single cluster